Use do/while for error macros instead of if/else/void.

tdaven opened the pull request:
Using do/while instead of the if/else/void pattern in the error macros is
a common way to require a semi-colon after that macro that also avoids the
extra else branch.  This fixes the issues with debug builds with gcc 12
which warn for the else branch that lacks a return.

Fixes #58747

timothyqiu commented on core/error/error_macros.h:
```diff
@@ -487,34 +515,37 @@ void _err_flush_stdout();
+	do {                                                                                                               \
+		if (unlikely(m_cond)) {                                                                                        \
+			_err_print_error(FUNCTION_STR, __FILE__, __LINE__, "Condition \"" _STR(m_cond) "\" is true. Continuing."); \
+			continue;                                                                                                  \
```
This changes how `continue` work here.

tdaven replied:
Yeah, the break/continue macros are fine as is so I have reverted those.

akien-mga replied:
Seems like you only reverted one of the three, and it's now broken.

tdaven replied:
Apparently I forgot to add my latest edits.  The intent was all the break/continue version stay as they are now using ((void)0) since we don't want to change the behavior of the macros used inside loops.  Macros that only return or generate a trap can be wrapped.  I'll rebase and push again after a clean build.

tdaven committed:
Use do/while for some error macros instead of if/else/void.

Using do/while instead of the if/else/void pattern in the error macros is
a common way to require a semi-colon after the macro that also avoids
the extra else branch. This fixes  issues with debug builds with gcc 12
which warns for the else branch that lacks a return.

Fixes #58747

Fix trailing whitespace and improve comment spacing.

Co-authored-by: Rémi Verschelde <rverschelde@gmail.com>

akien-mga commented on core/error/error_macros.h:
```diff
@@ -100,8 +100,8 @@ void _err_flush_stdout();
  *
- * The `((void)0)` no-op statement is used as a trick to force us to put a semicolon after
- * those macros, making them look like proper statements.
+ * The `((void)0)` no-op and `do{...} while(false)` statements are used as a trick to force 
```
There's a trailing space on this line. And might be good to add spaces in the quoted pseudo code to follow our convention for readability:
```suggestion
 * The `((void)0)` no-op and `do {...} while (false)` statements are used as a trick to force
```

akien-mga commented:
I remembered that we used to use `do {...} while (0)` for these and it was changed in #36011, due to the issue with `break;` and `continue;`. This PR now keeps the `if () else ((void)0)` for `ERR_CONTINUE*` and `ERR_BREAK*` macros so it shouldn't reintroduce the bug. But we have to hope that GCC 12 doesn't trigger its false positive on those macros. CC @madmiraal @Faless 

I also wondered if there's any difference in terms of how the compiler will optimize these, but it seems to give the same assembly for latest MSVC, GCC and Clang with either version: https://godbolt.org/z/bTqjM7Tob (not sure that my test really puts the compiler in a situation where it might have to do other choices).

akien-mga commented:
After discussing with @Faless, we agreed that this solution isn't really satisfactory as the GCC false positive could still happen with uses of `ERR_BREAK*` and `ERR_CONTINUE*` (it currently doesn't, but could if we were to use a construct similar to https://gcc.gnu.org/bugzilla/show_bug.cgi?id=104787#c1, which means simply creating an instance of an object with a destructor before the guaranteed return branch).

And modifying a version which was found to be better (since it doesn't have the `break;`/`continue;` issue) to work around a confirmed upstream compiler bug is something we prefer not to do.

So instead I made a PR to disable `-Werror=return-type` (which used to always be enabled, even when not treating warnings as errors) when using GCC 12+: #60652.

Thanks for your contribution nevertheless!
